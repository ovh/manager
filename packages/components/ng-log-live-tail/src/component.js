import template from './template.html';
import controller from './controller';

export default {
  controller,
  template,
  transclude: {
    tile: '?tile',
    description: '?description',
  },
  bindings: {
    logApiUrl: '<',
    logKind: '<',
    logKeys: '<',
  },
};
