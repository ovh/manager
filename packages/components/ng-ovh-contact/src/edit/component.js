import controller from './controller';
import template from './template.html';

export default {
  require: {
    ovhContactCtrl: '^ovhContact',
  },
  template,
  controller,
  bindings: {
    options: '=?ovhContactChoiceOptions',
    customList: '=?ovhContactChoiceCustomList',
  },
};
