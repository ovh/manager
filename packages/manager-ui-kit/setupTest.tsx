import * as React from 'react';
import { ComponentType } from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';
import i18n from 'i18next';
import {
  RenderOptions,
  RenderResult,
  render,
  cleanup,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import 'element-internals-polyfill';
import { vi } from 'vitest';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import ActionMenuTransFR from './src/components/action-menu/translations/Messages_fr_FR.json';
import DatagridTransFR from './src/components/datagrid/translations/Messages_fr_FR.json';
import FiltersTransFR from './src/components/filters/translations/Messages_fr_FR.json';

i18n.use(initReactI18next).init({
  fallbackLng: 'fr_FR',
  interpolation: {
    escapeValue: false,
  },
  resources: {
    fr: {
      'action-menu': {
        ...ActionMenuTransFR,
      },
      datagrid: {
        ...DatagridTransFR,
      },
      filterAdd: {
        ...FiltersTransFR,
      },
    },
  },
});

export const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      retry: false,
    },
    queries: {
      retry: false,
    },
  },
});

const Wrappers = ({ children }: { children: React.ReactElement }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: React.JSX.Element,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult =>
  render(ui, { wrapper: Wrappers as ComponentType, ...options });

export { customRender as render, cleanup };

// This polyfill exists because of an issue with jsdom and the EventTarget class
// when testing a component with an OdsDatepicker (addEventListener crashes at component initialization).
// Fix from issue: https://github.com/jsdom/jsdom/issues/2156
global.EventTarget = class {
  listeners = {};

  addEventListener(type, listener) {
    this.listeners = this.listeners || {};
    (this.listeners[type] || (this.listeners[type] = new Set())).add(listener);
  }

  removeEventListener(type, listener) {
    if (this.listeners && this.listeners[type]) {
      this.listeners[type].delete(listener);
    }
  }

  dispatchEvent(event) {
    this.listeners[event.type].forEach((listener) => listener(event));
    return !event.defaultPrevented;
  }
};

const ResizeObserverMock = vi.fn((callback) => {
  // Create a mock ResizeObserverEntry with the expected structure
  const mockEntry = {
    target: document.createElement('div'),
    contentRect: {
      width: 100,
      height: 100,
      top: 0,
      left: 0,
      bottom: 100,
      right: 100,
      x: 0,
      y: 0,
    },
    borderBoxSize: [{ width: 100, height: 100 }],
    contentBoxSize: [{ width: 100, height: 100 }],
    devicePixelContentBoxSize: [{ width: 100, height: 100 }],
  };

  // Call the callback with an array of entries as the real ResizeObserver does
  callback([mockEntry]);

  return {
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
  };
});

vi.stubGlobal('ResizeObserver', ResizeObserverMock);

// Mock getBoundingClientRect for virtualization and components
Element.prototype.getBoundingClientRect = vi.fn(() => ({
  width: 1000,
  height: 600,
  top: 0,
  left: 0,
  bottom: 600,
  right: 1000,
  x: 0,
  y: 0,
  toJSON: () => {},
}));

// Mock HTMLElement methods for components
HTMLElement.prototype.scrollTo = vi.fn();
HTMLElement.prototype.scrollIntoView = vi.fn();

// Mock scroll properties needed by ODS Modal
Object.defineProperty(Element.prototype, 'scrollLeft', {
  configurable: true,
  writable: true,
  value: 0,
});

Object.defineProperty(Element.prototype, 'scrollTop', {
  configurable: true,
  writable: true,
  value: 0,
});

Object.defineProperty(Element.prototype, 'scrollWidth', {
  configurable: true,
  value: 0,
});

Object.defineProperty(Element.prototype, 'scrollHeight', {
  configurable: true,
  value: 0,
});

// Suppress unhandled async errors from ODS Modal cleanup
const originalConsoleError = console.error;
console.error = (...args: any[]) => {
  const errorMessage = args[0]?.toString() || '';
  // Suppress specific ODS Modal/Zag.js cleanup errors
  if (
    errorMessage.includes("reading 'left'") ||
    errorMessage.includes('[@zag-js/dismissable]') ||
    errorMessage.includes('preventScroll') ||
    errorMessage.includes('Uncaught Exception')
  ) {
    return;
  }
  originalConsoleError(...args);
};

// Catch unhandled errors from ODS Modal async cleanup
if (typeof window !== 'undefined') {
  window.addEventListener('error', (event) => {
    const errorMessage = event.message || event.error?.message || '';
    // Suppress ODS Modal cleanup errors
    if (
      errorMessage.includes("reading 'left'") ||
      errorMessage.includes('[@zag-js/dismissable]')
    ) {
      event.preventDefault();
      event.stopPropagation();
      return false;
    }
    return true;
  });

  window.addEventListener('unhandledrejection', (event) => {
    const errorMessage = event.reason?.message || '';
    // Suppress ODS Modal cleanup errors
    if (
      errorMessage.includes("reading 'left'") ||
      errorMessage.includes('[@zag-js/dismissable]')
    ) {
      event.preventDefault();
      event.stopPropagation();
    }
  });
}
