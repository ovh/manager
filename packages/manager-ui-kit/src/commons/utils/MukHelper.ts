export function hashCode(value: unknown): number {
  const str = (() => {
    switch (typeof value) {
      case 'number':
      case 'bigint':
      case 'boolean':
      case 'string':
        return String(value);
      case 'object':
        return value === null ? 'null' : JSON.stringify(value);
      case 'function':
        return 'function';
      case 'undefined':
        return 'undefined';
      default:
        return 'symbol';
    }
  })();

  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = (hash * 32 - hash + str.charCodeAt(i)) % 2147483647;
  }

  return hash;
}
