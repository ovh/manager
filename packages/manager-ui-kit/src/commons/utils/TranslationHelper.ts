import i18next from 'i18next';
import type { Resource, ResourceLanguage } from 'i18next';

import { locales } from '@/commons/settings/Locales.constants';

/**
 * Manager Fallback Language
 */
const FALLBACK_LANGUAGE = 'fr_FR';

/**
 * Normalizes a language code to a format compatible with i18next loaders.
 * Converts hyphenated codes to underscore format and ensures region codes exist.
 *
 * Examples:
 * - 'fr-FR' → 'fr_FR'
 * - 'en' → 'en_GB'
 * - 'de' → 'de_DE'
 *
 * @param language - The language code from browser or user input.
 * @returns Normalized language code in the format `ll_RR` (e.g., 'en_GB').
 */
export const normalizeLanguageCode = (language: string): string => {
  let normalizedLang = language.replace('-', '_');

  if (!normalizedLang.includes('_')) {
    normalizedLang =
      normalizedLang === 'en' ? 'en_GB' : `${normalizedLang}_${normalizedLang.toUpperCase()}`;
  }

  return normalizedLang;
};

/**
 * Builds a resource loader function for dynamic translation loading.
 *
 * @param translationLoaders - A map of language codes to dynamic import functions for translations.
 * @param namespace - The namespace under which translations will be added.
 */
export const buildTranslationResources =
  (
    translationLoaders: Record<
      string,
      () => Promise<{ default?: Resource; [key: string]: unknown }>
    >,
    namespace: string,
  ) =>
  async (language: string): Promise<boolean> => {
    const normalizedLang = normalizeLanguageCode(language);

    // Always load fallbackLang once, before the user language
    if (!i18next.hasResourceBundle(FALLBACK_LANGUAGE, namespace)) {
      const fallbackLoader = translationLoaders[FALLBACK_LANGUAGE];
      if (fallbackLoader) {
        try {
          console.info(
            `Loading fallback language: ${FALLBACK_LANGUAGE} for namespace: ${namespace}`,
          );
          const fallbackModule = await fallbackLoader();
          const resources = (fallbackModule.default ?? fallbackModule) as Resource;
          i18next.addResources(FALLBACK_LANGUAGE, namespace, resources);
        } catch (error) {
          console.error(`Failed to load fallback translations (${FALLBACK_LANGUAGE}):`, error);
        }
      }
    }

    // Then load the requested language
    if (
      normalizedLang !== FALLBACK_LANGUAGE &&
      !i18next.hasResourceBundle(normalizedLang, namespace)
    ) {
      const loader = translationLoaders[normalizedLang];
      if (loader) {
        try {
          const module = await loader();
          const resources = (module.default ?? module) as Resource;
          i18next.addResources(normalizedLang, namespace, resources);
        } catch (error) {
          console.warn(
            `Could not load ${normalizedLang}. Falling back to ${FALLBACK_LANGUAGE}.`,
            error,
          );
        }
      }
    }

    return true;
  };

/**
 * Initializes and manages i18next language updates and resource loading.
 * Listens to i18next events and browser language changes to update translations dynamically.
 *
 * @param translationLoaders - A map of language codes to dynamic import functions.
 * @param namespace - The namespace to be used for loading translations.
 */
export const buildTranslationManager = (
  translationLoaders: Record<string, () => Promise<{ default?: Resource; [key: string]: unknown }>>,
  namespace: string,
): void => {
  const loadTranslations = buildTranslationResources(translationLoaders, namespace);

  const handleLanguageChange = async (lang: string): Promise<void> => {
    const normalizedLang = normalizeLanguageCode(lang);
    await loadTranslations(normalizedLang);
    console.log('Language changed to:', normalizedLang);
  };

  // Ensure promises are properly handled for event-based hooks
  if (i18next.isInitialized) {
    void handleLanguageChange(i18next.language);
  } else {
    i18next.on('initialized', () => {
      void handleLanguageChange(i18next.language);
    });
  }

  i18next.on('languageChanged', (lang: string) => {
    void handleLanguageChange(lang);
  });
};

/**
 * Union type of supported locale identifiers.
 */
export type LocaleType = (typeof locales)[number];

/**
 * Asynchronous loader that returns an i18next-compatible language resource.
 */
export type TranslationLoaderType = () => Promise<ResourceLanguage>;

/**
 * Creates a resolver function that unwraps a dynamically imported module’s default export.
 *
 * @template T - Type of the resolved module content
 * @returns A function that converts `import('./file.json')` to a typed Promise<T>.
 *
 * @example
 * const resolveTranslationModule = createImportResolver<ResourceLanguage>();
 * const resource = await resolveTranslationModule(import('./Messages_fr_FR.json'));
 */
const createImportResolver =
  <T>() =>
  (p: Promise<{ default: unknown }>) =>
    p.then((m) => m.default as T);

/**
 * A specialized resolver for translation JSON modules,
 * producing i18next {@link ResourceLanguage} objects.
 */
export const resolveTranslationModule = createImportResolver<ResourceLanguage>();
