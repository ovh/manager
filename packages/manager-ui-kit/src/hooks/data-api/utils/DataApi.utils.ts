import type { ColumnSort } from '@tanstack/react-table';

import type { Filter } from '@ovh-ux/manager-core-api';

export const filterEquals = (a: Filter, b: Filter) =>
  a.key === b.key && a.value === b.value && a.comparator === b.comparator;

export const getSearchParamsObject = (search: URLSearchParams) =>
  Object.fromEntries([...search.entries()]);

export const parseSorting = (params: URLSearchParams, defaultSorting?: ColumnSort): ColumnSort => {
  const sorting: ColumnSort = {
    id: '',
    desc: false,
  };
  if (params.has('sort')) {
    const sortParam = params.get('sort');
    if (sortParam) {
      sorting.id = sortParam;
      if (params.has('sortOrder')) {
        sorting.desc = params.get('sortOrder') === 'desc';
      }
    }
  } else if (defaultSorting) {
    return defaultSorting;
  }
  return sorting;
};
