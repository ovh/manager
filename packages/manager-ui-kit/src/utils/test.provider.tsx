import React, { ComponentType } from 'react';
import { I18nextProvider, initReactI18next } from 'react-i18next';
import i18n from 'i18next';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import '@testing-library/jest-dom';
import 'element-internals-polyfill';

i18n.use(initReactI18next).init({
  fallbackLng: 'fr_FR',
  interpolation: {
    escapeValue: false,
  },
  resources: {
    en: {
      'action-menu': {
        common_actions: 'Actions',
      },
    },
  },
});

export const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      retry: false,
    },
    queries: {
      retry: false,
    },
  },
});

const Wrappers = ({ children }: { children: React.ReactElement }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: React.JSX.Element,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult =>
  render(ui, { wrapper: Wrappers as ComponentType, ...options });

export { customRender as render };
