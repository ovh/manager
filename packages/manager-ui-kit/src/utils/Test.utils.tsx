import { ReactNode } from 'react';

import { ShellContext, ShellContextType } from '@ovh-ux/manager-react-shell-client';

import { render } from '../../setupTest';

export const mockTrackPage = vitest.fn();
export const mockGetEnvironment = vitest.fn();

export const shellContext = {
  shell: {
    environment: {
      getEnvironment: mockGetEnvironment,
    },
    tracking: {
      trackPage: mockTrackPage,
    },
    navigation: {
      navigateTo: vitest.fn(),
      reload: vitest.fn(),
    },
    ux: {
      hidePreloader: vitest.fn(),
    },
  },
};

export const renderWithContext = ({ children }: { children: ReactNode }) =>
  render(
    <ShellContext.Provider value={shellContext as unknown as ShellContextType}>
      {children}
    </ShellContext.Provider>,
  );
