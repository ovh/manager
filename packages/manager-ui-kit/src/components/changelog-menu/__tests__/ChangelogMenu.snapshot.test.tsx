import { vitest } from 'vitest';

import { render } from '@/setupTest';

import { ChangelogMenu } from '../ChangelogMenu.component';
import { Links, chapters } from './ChangelogMenu.utils';

vitest.mock('@ovh-ux/manager-react-shell-client', () => ({
  useOvhTracking: vitest.fn(() => ({
    trackClick: vitest.fn(),
  })),
}));

describe('ChangelogMenu Snapshot Tests', () => {
  it('should match snapshot with default props', () => {
    const { container } = render(<ChangelogMenu links={Links} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot with chapters prop', () => {
    const { container } = render(<ChangelogMenu links={Links} chapters={chapters} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot with custom prefixes', () => {
    const customPrefixes = ['custom-prefix-1', 'custom-prefix-2'];
    const { container } = render(
      <ChangelogMenu links={Links} chapters={chapters} prefixes={customPrefixes} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot with different link URLs', () => {
    const customLinks = {
      changelog: 'https://custom-changelog.com',
      roadmap: 'https://custom-roadmap.com',
      'feature-request': 'https://custom-feature-request.com',
    };
    const { container } = render(<ChangelogMenu links={customLinks} chapters={chapters} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot with single chapter', () => {
    const { container } = render(<ChangelogMenu links={Links} chapters={['single-chapter']} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot with long URLs', () => {
    const longUrls = {
      changelog:
        'https://very-long-url-that-might-affect-rendering.com/path/to/changelog?param1=value1&param2=value2&param3=value3',
      roadmap:
        'https://very-long-url-that-might-affect-rendering.com/path/to/roadmap?param1=value1&param2=value2&param3=value3',
      'feature-request':
        'https://very-long-url-that-might-affect-rendering.com/path/to/feature-request?param1=value1&param2=value2&param3=value3',
    };
    const { container } = render(<ChangelogMenu links={longUrls} chapters={chapters} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
