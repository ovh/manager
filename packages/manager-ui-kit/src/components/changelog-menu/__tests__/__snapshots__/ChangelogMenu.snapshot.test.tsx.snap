// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChangelogMenu Snapshot Tests > should match snapshot with chapters prop 1`] = `
<button
  aria-controls=":r3:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--sm_1sbbz_34 _button--ghost_1sbbz_150 _popover-trigger_a014q_2"
  data-ods="popover-trigger"
  data-part="trigger"
  data-scope="popover"
  data-state="closed"
  dir="ltr"
  id="popover::r2::trigger"
  role="button"
  tabindex="0"
  type="button"
>
  Roadmap & Changelog
</button>
`;

exports[`ChangelogMenu Snapshot Tests > should match snapshot with custom prefixes 1`] = `
<button
  aria-controls=":r5:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--sm_1sbbz_34 _button--ghost_1sbbz_150 _popover-trigger_a014q_2"
  data-ods="popover-trigger"
  data-part="trigger"
  data-scope="popover"
  data-state="closed"
  dir="ltr"
  id="popover::r4::trigger"
  role="button"
  tabindex="0"
  type="button"
>
  Roadmap & Changelog
</button>
`;

exports[`ChangelogMenu Snapshot Tests > should match snapshot with default props 1`] = `
<button
  aria-controls=":r1:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--sm_1sbbz_34 _button--ghost_1sbbz_150 _popover-trigger_a014q_2"
  data-ods="popover-trigger"
  data-part="trigger"
  data-scope="popover"
  data-state="closed"
  dir="ltr"
  id="popover::r0::trigger"
  role="button"
  tabindex="0"
  type="button"
>
  Roadmap & Changelog
</button>
`;

exports[`ChangelogMenu Snapshot Tests > should match snapshot with different link URLs 1`] = `
<button
  aria-controls=":r7:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--sm_1sbbz_34 _button--ghost_1sbbz_150 _popover-trigger_a014q_2"
  data-ods="popover-trigger"
  data-part="trigger"
  data-scope="popover"
  data-state="closed"
  dir="ltr"
  id="popover::r6::trigger"
  role="button"
  tabindex="0"
  type="button"
>
  Roadmap & Changelog
</button>
`;

exports[`ChangelogMenu Snapshot Tests > should match snapshot with long URLs 1`] = `
<button
  aria-controls=":rb:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--sm_1sbbz_34 _button--ghost_1sbbz_150 _popover-trigger_a014q_2"
  data-ods="popover-trigger"
  data-part="trigger"
  data-scope="popover"
  data-state="closed"
  dir="ltr"
  id="popover::ra::trigger"
  role="button"
  tabindex="0"
  type="button"
>
  Roadmap & Changelog
</button>
`;

exports[`ChangelogMenu Snapshot Tests > should match snapshot with single chapter 1`] = `
<button
  aria-controls=":r9:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="_button_1sbbz_2 _button--primary_1sbbz_99 _button--sm_1sbbz_34 _button--ghost_1sbbz_150 _popover-trigger_a014q_2"
  data-ods="popover-trigger"
  data-part="trigger"
  data-scope="popover"
  data-state="closed"
  dir="ltr"
  id="popover::r8::trigger"
  role="button"
  tabindex="0"
  type="button"
>
  Roadmap & Changelog
</button>
`;
